tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/input_inline_fusion_pass.cc

```diff
@@ -0,0 +1,390 @@
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/LoopFusionUtils.h"
+#include "mlir/Transforms/LoopUtils.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/Utils.h"
+
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Location.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/lhlo_ops.h"
+#include "tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/transforms/lhlo_elemental_utils.h"
+#include "tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/transforms/map_lmhlo_to_scalar_op.h"
+
+using mlir::memref::LoadOp;
+
+namespace mlir {
+namespace lmhlo {
+
+#define GEN_PASS_CLASSES
+#include "mlir-hlo/Dialect/mhlo/transforms/lmhlo_passes.h.inc"
+
+namespace {
+
+// TODO: Maybe it worth explicitly adding the I/O buffers onto the outlining of
+// lmhlo::FusionOp and then mark IsolatedFromAbove for lmhlo::FusionOp. By this
+// way the fusion codegen passes can be OperationPass on lmhlo::FusionOp for
+// better compilation overhead.
+struct InputInlineFusion : public InputInlineFusionPassBase<InputInlineFusion> {
+  void runOnFunction() override;
+};
+
+}  // end anonymous namespace
+
+std::unique_ptr<FunctionPass> createInputInlineFusionPass() {
+  return std::make_unique<InputInlineFusion>();
+}
+
+namespace {
+
+constexpr unsigned c_MAX_ITERATION = 4096;
+
+// This pass works after LhloLegalizeRootsToParallelLoops pass for the
+// XLA-style fusion codegen. Refer to:
+// http://pai-blade.cn-hangzhou.oss.aliyun-inc.com/disc/fusion_codegen.png.
+//
+// It iteratively looks for the Lhlo op which is the direct producer of the
+// nested loops, and then inline fuse it if the fusion will not form a loop.
+//
+// The inline fusion action can be generalized as:
+// step 1: replace the producer Lhlo op into associate std op inside the nested
+// loops. step 2: remove the original Load ops inside the loops and insert new
+// Load ops.
+//
+// If there are multiple LoadOps with the same indices, they will be replaced
+// with the same op. This obtains the similar result as GeneratedValueCache.
+//
+// IRs after LhloLegalizeRootsToParallelLoops:
+//    "lmhlo.fusion"() ( {
+//       lmhlo.aaa(%0, %1, %2)
+//       lmhlo.bbb(%2, %3, %4)
+//       scf.parallel (...) {
+//          memref.load %4[...]
+//          ...
+//          memref.store ...
+//       }
+//    })
+//
+// IRs after one round of InputInlineFusionPattern:
+//    "lmhlo.fusion"() ( {
+//       lmhlo.aaa(%0, %1, %2)
+//       scf.parallel (...) {
+//          memref.load %2[...]
+//          ...
+//          memref.store ...
+//       }
+//    })
+//
+// Final IRs after this pass:
+//    "lmhlo.fusion"() ( {
```

Comment:
nit: do we need this `lmhlo.fusion` at this point?

Reply:
Could you explain a little bit more? I didn't get it.

Reply:
If you refer to the question of "can we do fusion codegen without the fusion pass?".
I think we still need it to identify the root ops and boundaries, and potentially evolve to some more elegant fusion strategies.

Reply:
since the lmhlo fusion only contains a single scf.parallel, the question may be what is the fusion encapsulation providing? Why can't you eliminate the fusion and just have the parallel loop by itself?

Reply:
The two codegen passes in this PR still need to know the boundaries of a kernel, which is provided with lmhlo.fusion.
After its body is fully lowered, we'll eliminate the fusion and inline its body contents into the upper level region. We'll do this with another FusionInliner pass, which has not been submitted yet.